File: packages/validator-php/src/V2/Compose/Compose.php
<?php

declare(strict_types=1);

namespace FormSpec\Validator\V2\Compose;

/**
 * Composition orchestrator (SPEC-V2 §5, G5) — the parser's FIRST pass.
 * Byte-for-byte port of validator-ts/src/v2/compose/compose.ts.
 *
 * Resolution order (SPEC-V2 §5, verbatim):
 *   (1) $ref   — expand file/path base, recursively (nested $ref included) into a
 *                single properties base.
 *   (2) $patch — overlay add/remove/replace and deep-path set on that base.
 *   (3) the result is the equivalent SINGLE SPEC (composition keys eliminated).
 *   (4) the field layer (validate/design/behavior/options) then applies to it.
 *
 * G5 (SPEC §2): the parser expands composition first, producing a single spec,
 * BEFORE applying the field layer — without composition a $ref-using spec cannot
 * even be loaded. So composition is a pre-processing pass that runs BEFORE
 * validation/render, not a validation step.
 *
 * v1's positional array_merge priority is normalized here: $ref = base (first),
 * $patch = overlay (later) — base is laid down, patch overrides.
 *
 * composeProperties is the entry point: it operates on a properties map (the
 * composition entry point, SPEC §2) where $ref/$patch may sit alongside named
 * child fields. composeSpec recurses the whole field tree so a $ref nested inside
 * any child properties is expanded too. Both throw a ComposeLoadError for any
 * unresolved composition (never valid:true).
 */
final class Compose
{
    /** Sentinel for "no $patch seen" (distinct from a present null $patch). */
    private const UNDEFINED = "\0__compose_undefined__\0";

    /**
     * Compose a properties map: expand $ref to a base, overlay $patch, return the
     * single (composition-free) properties map. Named sibling keys follow v1
     * declaration order — a key declared after $ref overrides the base; a key
     * declared before it is overridden by the base.
     *
     * @param array<string, mixed> $properties
     * @return array<string, mixed>
     */
    public static function properties(
        array $properties,
        FileLoader $loader,
        string $basepath = '',
    ): array {
        $base = [];
        $patch = self::UNDEFINED;
        $own = [];

        foreach ($properties as $k => $v) {
            if ($k === '$ref') {
                // $ref array_merges onto whatever was declared before it (v1 order).
                $base = self::merge($own, Ref::resolve($v, $basepath, $loader));
                $own = [];
            } elseif ($k === '$patch') {
                $patch = $v;
            } else {
                $own[$k] = $v;
            }
        }

        // No composition keys: still recurse into children so nested $ref expands.
        $result = self::merge($base, $own);
        if ($patch !== self::UNDEFINED) {
            $result = Patch::apply($result, $patch);
        }

        // Recurse into every child field's properties (the tree may compose deeper).
        foreach ($result as $fieldName => $field) {
            if (self::isPlainObject($field)) {
                /** @var array<string, mixed> $field */
                $result[$fieldName] = self::spec($field, $loader, $basepath);
            }
        }

        return $result;
    }

    /**
     * Compose a full field spec: expand a field-level $ref/$patch, then recurse
     * into its properties (which may itself compose). Returns the single spec.
     *
     * @param array<string, mixed> $spec
     * @return array<string, mixed>
     */
    public static function spec(
        array $spec,
        FileLoader $loader,
        string $basepath = '',
    ): array {
        // Field-level $ref / $patch (a field may inherit a whole base spec).
        if (\array_key_exists('$ref', $spec) || \array_key_exists('$patch', $spec)) {
            $base = [];
            $patch = self::UNDEFINED;
            $own = [];
            foreach ($spec as $k => $v) {
                if ($k === '$ref') {
                    // Field-level $ref resolves a file's properties layer too
                    // (v1 detectKey).
                    $base = self::merge($own, Ref::resolve($v, $basepath, $loader));
                    $own = [];
                } elseif ($k === '$patch') {
                    $patch = $v;
                } else {
                    $own[$k] = $v;
                }
            }
            $resolved = self::merge($base, $own);
            if ($patch !== self::UNDEFINED) {
                $resolved = Patch::apply($resolved, $patch);
            }
        } else {
            $resolved = $spec;
        }

        // Recurse into properties (composition entry point, SPEC §2).
        if (\array_key_exists('properties', $resolved) && self::isPlainObject($resolved['properties'])) {
            /** @var array<string, mixed> $props */
            $props = $resolved['properties'];
            $resolved['properties'] = self::properties($props, $loader, $basepath);
        }

        return $resolved;
    }

    /**
     * Shallow merge { ...a, ...b }: b overrides a on clash, b's new keys append.
     *
     * @param array<string, mixed> $a
     * @param array<string, mixed> $b
     * @return array<string, mixed>
     */
    private static function merge(array $a, array $b): array
    {
        $out = $a;
        foreach ($b as $k => $v) {
            $out[$k] = $v;
        }
        return $out;
    }

    /**
     * JS "plain object" predicate: non-list array, or empty array (read as {});
     * never a scalar/null/list.
     */
    private static function isPlainObject(mixed $v): bool
    {
        if (!\is_array($v)) {
            return false;
        }
        if ($v === []) {
            return true;
        }
        return !\array_is_list($v);
    }
}

File: packages/validator-rust/src/compose/compose.rs
//! Composition orchestrator (SPEC §5, G5) — the parser's FIRST pass.
//!
//! Resolution order (SPEC §5, verbatim):
//!   (1) $ref   — expand file/path base, recursively (nested $ref included) into a
//!                single properties base.
//!   (2) $patch — overlay add/remove/replace and deep-path set on that base.
//!   (3) the result is the equivalent SINGLE SPEC (composition keys eliminated).
//!   (4) the field layer (validate/design/behavior/options) then applies to it.
//!
//! G5 (SPEC §2): the parser expands composition first, producing a single spec,
//! BEFORE applying the field layer — without composition a `$ref`-using spec
//! cannot even be loaded. So composition is a pre-processing pass that runs
//! BEFORE validation/render, not a validation step.
//!
//! legacy's positional `array_merge` priority is normalized here: `$ref` = base
//! (first), `$patch` = overlay (later) — base is laid down, patch overrides.
//!
//! `compose_properties` is the entry point: it operates on a `properties` map (the
//! composition entry point, SPEC §2 / types.rs:73) where `$ref`/`$patch` may sit
//! alongside named child fields. `compose_spec` recurses the whole field tree so a
//! `$ref` nested inside any child `properties` is expanded too. Both return a
//! `ComposeLoadError` for any unresolved composition (never `valid:true`).

use std::collections::BTreeSet;

use serde_json::{Map, Value};

use super::errors::ComposeResult;
use super::loader::FileLoader;
use super::member_order::member_ordered_map;
use super::patch::apply_patch;
use super::ref_::resolve_ref;

/// Options for a composition pass.
#[derive(Debug, Clone, Default)]
pub struct ComposeOptions {
    /// Basepath for relative `$ref` resolution (legacy ReferenceResolver basepath).
    pub basepath: String,
}

impl ComposeOptions {
    /// Construct options with the given basepath.
    pub fn with_basepath(basepath: impl Into<String>) -> Self {
        ComposeOptions {
            basepath: basepath.into(),
        }
    }
}

/// Compose a `properties` map: expand `$ref` to a base, overlay `$patch`, return
/// the single (composition-free) properties map. Named sibling keys follow
/// declaration order — a key declared after `$ref` overrides the base; a key
/// declared before it is overridden by the base. The input and the result are in
/// specification member order at every depth.
pub fn compose_properties(
    properties: Map<String, Value>,
    loader: &dyn FileLoader,
    opts: &ComposeOptions,
) -> ComposeResult<Map<String, Value>> {
    let composed = expand_properties(member_ordered_map(&properties), loader, opts)?;
    Ok(member_ordered_map(&composed))
}

/// Compose a full field spec: expand a field-level `$ref`/`$patch`, then recurse
/// into its `properties` (which may itself compose). Returns the single spec. The
/// input and the result are in specification member order at every depth.
pub fn compose_spec(
    spec: Map<String, Value>,
    loader: &dyn FileLoader,
    opts: &ComposeOptions,
) -> ComposeResult<Map<String, Value>> {
    let composed = expand_spec(member_ordered_map(&spec), loader, opts)?;
    Ok(member_ordered_map(&composed))
}

/// `compose_properties` over a member-ordered input, without ordering the result.
fn expand_properties(
    properties: Map<String, Value>,
    loader: &dyn FileLoader,
    opts: &ComposeOptions,
) -> ComposeResult<Map<String, Value>> {
    let basepath = &opts.basepath;
    let visiting: BTreeSet<String> = BTreeSet::new();

    let mut base: Map<String, Value> = Map::new();
    let mut patch: Option<Value> = None;
    let mut own: Map<String, Value> = Map::new();

    for (k, v) in properties {
        if k == "$ref" {
            // $ref array_merges onto whatever was declared before it (legacy order).
            let resolved = resolve_ref(&v, basepath, loader, &visiting)?;
            let mut next_base: Map<String, Value> = Map::new();
            for (ok, ov) in own.iter() {
                next_base.insert(ok.clone(), ov.clone());
            }
            for (rk, rv) in resolved {
                next_base.insert(rk, rv);
            }
            base = next_base;
            own.clear();
        } else if k == "$patch" {
            patch = Some(v);
        } else {
            own.insert(k, v);
        }
    }

    // No composition keys: still recurse into children so nested $ref expands.
    let mut result = base;
    for (k, v) in own {
        result.insert(k, v);
    }
    if let Some(p) = patch {
        result = apply_patch(result, &p)?;
    }

    // Recurse into every child field's `properties` (the tree may compose deeper).
    // ORDER: clone the child value out (get().cloned()), compose it, then re-insert
    // under the SAME key. IndexMap::insert keeps an existing key in its place — so
    // declaration order is preserved. NEVER remove()+insert (remove = swap_remove,
    // which moves the last key into this slot and corrupts the positional order
    // that legacy's array_merge merge priority is load-bearing on).
    let keys: Vec<String> = result.keys().cloned().collect();
    for field_name in keys {
        if let Some(Value::Object(child)) = result.get(&field_name).cloned() {
            let composed = expand_spec(child, loader, opts)?;
            result.insert(field_name, Value::Object(composed));
        }
    }

    Ok(result)
}

/// `compose_spec` over a member-ordered input, without ordering the result.
fn expand_spec(
    spec: Map<String, Value>,
    loader: &dyn FileLoader,
    opts: &ComposeOptions,
) -> ComposeResult<Map<String, Value>> {
    let basepath = &opts.basepath;
    let visiting: BTreeSet<String> = BTreeSet::new();

    let mut resolved: Map<String, Value>;

    // Field-level $ref / $patch (a field may inherit a whole base spec).
    if spec.contains_key("$ref") || spec.contains_key("$patch") {
        let mut base: Map<String, Value> = Map::new();
        let mut patch: Option<Value> = None;
        let mut own: Map<String, Value> = Map::new();

        for (k, v) in spec {
            if k == "$ref" {
                // Field-level $ref resolves a file's properties layer too (detectKey).
                let res = resolve_ref(&v, basepath, loader, &visiting)?;
                let mut next_base: Map<String, Value> = Map::new();
                for (ok, ov) in own.iter() {
                    next_base.insert(ok.clone(), ov.clone());
                }
                for (rk, rv) in res {
                    next_base.insert(rk, rv);
                }
                base = next_base;
                own.clear();
            } else if k == "$patch" {
                patch = Some(v);
            } else {
                own.insert(k, v);
            }
        }
        let mut merged = base;
        for (k, v) in own {
            merged.insert(k, v);
        }
        if let Some(p) = patch {
            merged = apply_patch(merged, &p)?;
        }
        resolved = merged;
    } else {
        resolved = spec;
    }

    // Recurse into `properties` (composition entry point, SPEC §2 / types.rs:73).
    // ORDER: clone out, compose, re-insert under the SAME key so `properties` keeps
    // its declared position (IndexMap::insert preserves an existing key's slot).
    // remove() here would swap_remove and shove `properties` to a wrong position.
    if let Some(Value::Object(props)) = resolved.get("properties").cloned() {
        let composed = expand_properties(props, loader, opts)?;
        resolved.insert("properties".to_string(), Value::Object(composed));
    }

    Ok(resolved)
}

#[cfg(test)]
mod order_tests {
    //! Verify the literal unsorted key sequence for the composition tree. Value
    //! equality does not detect reordering, so these tests compare keys directly.

    use super::{compose_properties, compose_spec, ComposeOptions};
    use crate::compose::loader::MemoryLoader;
    use serde_json::{json, Value};

    fn keys(v: &Value) -> Vec<String> {
        v.as_object().unwrap().keys().cloned().collect()
    }

    fn loader(files: Value) -> MemoryLoader {
        MemoryLoader::new(files.as_object().unwrap().clone())
    }

    fn props(v: Value) -> serde_json::Map<String, Value> {
        v.as_object().unwrap().clone()
    }

    #[test]
    fn child_recursion_preserves_field_declaration_order() {
        // Three sibling fields, each with a nested `properties` that composes. The
        // re-insert after composing each child must keep a,b,c in order — a
        // remove()+insert (swap_remove) would scramble them to e.g. c,b,a / a,c,b.
        let ml = loader(json!({}));
        let entry = props(json!({
            "a": {"type": "group", "properties": {"x": {"type": "text"}}},
            "b": {"type": "group", "properties": {"y": {"type": "text"}}},
            "c": {"type": "group", "properties": {"z": {"type": "text"}}}
        }));
        let out = compose_properties(entry, &ml, &ComposeOptions::default()).unwrap();
        assert_eq!(keys(&Value::Object(out)), vec!["a", "b", "c"]);
    }

    #[test]
    fn ref_base_then_sibling_after_keeps_base_order_and_appends_new() {
        // base = {a,b}; entry declares $ref then sibling c → a,b (base order) then c.
        let ml = loader(json!({
            "base.yml": {"properties": {"a": {"t": 1}, "b": {"t": 2}}}
        }));
        let entry = props(json!({"$ref": "base.yml", "c": {"t": 3}}));
        let out = compose_properties(entry, &ml, &ComposeOptions::default()).unwrap();
        assert_eq!(keys(&Value::Object(out)), vec!["a", "b", "c"]);
    }

    #[test]
    fn sibling_after_ref_overrides_base_in_its_base_slot() {
        // a declared AFTER $ref overrides the base value but the key keeps the
        // base's slot (insert on an existing key preserves position): a,b order.
        let ml = loader(json!({
            "base.yml": {"properties": {"a": {"from": "base"}, "b": {"t": 2}}}
        }));
        let entry = props(json!({"$ref": "base.yml", "a": {"from": "override"}}));
        let out = compose_properties(entry, &ml, &ComposeOptions::default()).unwrap();
        let out = Value::Object(out);
        assert_eq!(keys(&out), vec!["a", "b"]);
        assert_eq!(out["a"]["from"], json!("override"));
    }

    #[test]
    fn spec_level_properties_keeps_its_declared_position() {
        // type, properties, behavior — composing `properties` must NOT move it to
        // the end (it sits in slot 1 and must stay there).
        let ml = loader(json!({
            "group.yml": {"properties": {"child": {"type": "text"}}}
        }));
        let entry = props(json!({
            "type": "group",
            "properties": {"$ref": "group.yml"},
            "behavior": {"collapsible": true}
        }));
        let out = compose_spec(entry, &ml, &ComposeOptions::default()).unwrap();
        let out = Value::Object(out);
        assert_eq!(keys(&out), vec!["type", "properties", "behavior"]);
        assert_eq!(keys(&out["properties"]), vec!["child"]);
    }

    /// An object with members inserted in exactly the given order.
    fn written(members: &[(&str, Value)]) -> Value {
        let mut map = serde_json::Map::new();
        for (name, value) in members {
            map.insert((*name).to_string(), value.clone());
        }
        Value::Object(map)
    }

    #[test]
    fn patched_array_index_member_joins_in_member_order() {
        let base = written(&[("b", json!({"t": "b"})), ("a", json!({"t": "a"}))]);
        let ml = loader(json!({"base.yml": {"properties": base}}));
        let entry = props(json!({"$ref": "base.yml", "$patch": {"10": {"t": "ten"}}}));
        let out = compose_properties(entry, &ml, &ComposeOptions::default()).unwrap();
        assert_eq!(keys(&Value::Object(out)), vec!["10", "b", "a"]);
    }

    /// A loader that returns documents exactly as written, like a custom loader.
    struct Written(Value);

    impl crate::compose::FileLoader for Written {
        fn normalize(&self, path: &str, _basepath: &str) -> String {
            path.to_string()
        }
        fn load(
            &self,
            _key: &str,
        ) -> crate::compose::ComposeResult<serde_json::Map<String, Value>> {
            Ok(self.0.as_object().unwrap().clone())
        }
    }

    #[test]
    fn loaded_documents_are_read_in_member_order() {
        let doc = written(&[(
            "properties",
            written(&[("b", json!({})), ("10", json!({})), ("a", json!({}))]),
        )]);
        let custom = Written(json!({"properties": {"10": {"from": "base"}}}));
        let out = compose_properties(
            props(json!({"$ref": "doc.yml"})),
            &Written(doc),
            &ComposeOptions::default(),
        )
        .unwrap();
        assert_eq!(keys(&Value::Object(out)), vec!["10", "b", "a"]);
        // Written as $ref, 10: in member order 10 precedes $ref, so the base overrides it.
        let entry = written(&[("$ref", json!("x.yml")), ("10", json!({"from": "entry"}))]);
        let out = compose_properties(
            entry.as_object().unwrap().clone(),
            &custom,
            &ComposeOptions::default(),
        )
        .unwrap();
        assert_eq!(Value::Object(out)["10"]["from"], json!("base"));
    }

    #[test]
    fn validation_errors_follow_member_order_and_data_is_not_reordered() {
        use crate::validate::{validate, ValidateOptions};
        let required = json!({"type": "text", "validate": {"required": true}});
        let properties = written(&[
            ("b", required.clone()),
            ("10", required.clone()),
            ("a", required),
        ]);
        let data = written(&[("a", json!("")), ("b", json!(""))]);
        let result = validate(
            &json!({"type": "group", "properties": properties}),
            &data,
            &ValidateOptions::default(),
        )
        .unwrap();
        let paths: Vec<&str> = result.errors.iter().map(|e| e.path.as_str()).collect();
        assert_eq!(paths, vec!["10", "b", "a"]);
        assert_eq!(keys(&data), vec!["a", "b"]);
    }

    #[test]
    fn nested_ref_flatten_keeps_inner_then_own_order() {
        // base.yml: $ref inner.yml (deep) then own → inner base laid first, own
        // appended: deep, own (declaration order across the $ref boundary).
        let ml = loader(json!({
            "base.yml": {"properties": {"$ref": "inner.yml", "own": {"from": "base"}}},
            "inner.yml": {"properties": {"deep": {"from": "inner"}}}
        }));
        let entry = props(json!({"$ref": "base.yml"}));
        let out = compose_properties(entry, &ml, &ComposeOptions::default()).unwrap();
        assert_eq!(keys(&Value::Object(out)), vec!["deep", "own"]);
    }
}
